app/src/main/java/org/schabi/newpipe/player/ui/PopupPlayerUi.java

```diff
@@ -0,0 +1,485 @@
+            popupLayoutParams.y = screenHeight - popupLayoutParams.height;
+        }
+    }
+
+    public void updateScreenSize() {
+        final DisplayMetrics metrics = new DisplayMetrics();
+        windowManager.getDefaultDisplay().getMetrics(metrics);
+
+        screenWidth = metrics.widthPixels;
+        screenHeight = metrics.heightPixels;
+        if (DEBUG) {
+            Log.d(TAG, "updateScreenSize() called: screenWidth = ["
+                    + screenWidth + "], screenHeight = [" + screenHeight + "]");
+        }
+    }
+
+    /**
+     * Changes the size of the popup based on the width.
+     * @param width the new width, height is calculated with
+     *              {@link PlayerHelper#getMinimumVideoHeight(float)}
+     */
+    public void changePopupSize(final int width) {
+        if (DEBUG) {
+            Log.d(TAG, "changePopupSize() called with: width = [" + width + "]");
+        }
+
+        if (anyPopupViewIsNull()) {
+            return;
+        }
+
+        final float minimumWidth = context.getResources().getDimension(R.dimen.popup_minimum_width);
+        final int actualWidth = (int) (width > screenWidth ? screenWidth
+                : (width < minimumWidth ? minimumWidth : width));
+        final int actualHeight = (int) getMinimumVideoHeight(width);
+        if (DEBUG) {
+            Log.d(TAG, "updatePopupSize() updated values:"
+                    + "  width = [" + actualWidth + "], height = [" + actualHeight + "]");
+        }
+
+        popupLayoutParams.width = actualWidth;
+        popupLayoutParams.height = actualHeight;
+        binding.surfaceView.setHeights(popupLayoutParams.height, popupLayoutParams.height);
+        windowManager.updateViewLayout(binding.getRoot(), popupLayoutParams);
+    }
+
+    @Override
+    protected float calculateMaxEndScreenThumbnailHeight(@NonNull final Bitmap bitmap) {
+        // no need for the end screen thumbnail to be resized on popup player: it's only needed
+        // for the main player so that it is enlarged correctly inside the fragment
+        return bitmap.getHeight();
+    }
+    //endregion
+
+
+    /*//////////////////////////////////////////////////////////////////////////
+    // Popup closing
+    //////////////////////////////////////////////////////////////////////////*/
+    //region Popup closing
+
+    public void closePopup() {
+        if (DEBUG) {
+            Log.d(TAG, "closePopup() called, isPopupClosing = " + isPopupClosing);
+        }
+        if (isPopupClosing) {
+            return;
+        }
+        isPopupClosing = true;
+
+        player.saveStreamProgressState();
+        windowManager.removeView(binding.getRoot());
+
+        animatePopupOverlayAndFinishService();
+    }
+
+    public boolean isPopupClosing() {
+        return isPopupClosing;
+    }
+
+    public void removePopupFromView() {
+        // wrap in try-catch since it could sometimes generate errors randomly
```

Comment:
Are the random errors new or have they been there before?

If they were there before we should link the corresponding issue.
If they haven't we should maybe create a new issue.

Reply:
I will not take care of this in this PR as it's just copied code: https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/java/org/schabi/newpipe/player/Player.java#L1571